File: Scala-Multiproc/src/stock/Stock.scala
/*
*TITLE: 	Stock SCALA Sorting Application
*AUTOHR: 	THOMAS TRASK
*STARTDATE: 12-5-2012
*
*PURPOSE: 	TO GENERATE A LIST OF RANDOM NUMBERS AND
*			SORT THEM USING THE BUILT-IN Scala UTILITIES.  THIS APPLICATION WAS
*			DEVELOPED AS A "CONTROL" BENCHMARK TO DETERMINE IF USING MULTITHREADING
*			/CLUSTER COMPUTING HAS ANY REAL PERFORMANCE BENEFIT.
*
*PROCESS:	THIS PROGRAM CONSISTS OF 3 PARTS: GENERATION, SORT AND
*			VERIFICATION & OUTPUT.
*			GENERATION:	THE APPLICATION GENERETATES A SET OF RANDOM NUMBERS TO SORT.
*			SORT:		THE APPLICATION THEN SORTS THIS DATA SET USING
						THE "scala.util.Sorting.quickSort" FUNCTION.
*			VERIFICATION & OUTPUT: THE SORTED DATA SET IS VERIFIED AS WORKING.
*					ONCE VERIFIED, ALL IMPORTANT PARAMETERS ARE WRITTEN OUT TO
*					AN OUTPUT FILE.
*
*EXECUTION:	Once the scala application set ahs been compiled,
* 			The stock application can be executed usign the following command:
*    			scala stock.Stock -n <int> -o <somefile.ext>
*			
*
*INPUT PARAMETERS:
*			-n:		Number of Ints to generate and sort
*			-o:		The name of the output file to write the results to.				
 */

package stock

import optional.ParseParms

object Stock extends scala.App {

  override def main(args: Array[String]): Unit =
    {

      var size: Int = 100;
      var outputFileName = "scala.stock.csv"

      implicit def string2Int(s: String): Int = augmentString(s).toInt

      val helpString = "[-n size -o outputFile.csv]"
      val pp = new ParseParms(helpString)
      pp.parm("-n", size.toString())

      val result = pp.validate(args.toList)
      
      println(if (result._1) result._3 else result._2)

      val n: Int = string2Int(result._3("-n"))
      if (n > 0) {
        size = n;
      }

      println(if (result._1) result._3 else result._2)

      System.out.println("Generating Numbers Now!");

      val testList: Array[Int] = math.NumberGenerator.GenerateRandomNumberList(size, Int.MaxValue)

      System.out.println("Done Generating List!");
      System.out.println("Staritng Sort Now!");
      var startSortTime: java.util.Date = new (java.util.Date)

      //SORTS DATA
      val results = scala.util.Sorting.quickSort(testList)
      
      
      var endSortTime: java.util.Date = new (java.util.Date)
      var sortTime = endSortTime.getTime() - startSortTime.getTime();
      System.out.println("Done Sorting in:" + (sortTime).toString());

      System.out.println("Writing to output file:" + outputFileName);

      val fw = new java.io.FileWriter(outputFileName, true)
      try {
        var df: java.text.DateFormat = new java.text.SimpleDateFormat("MM/dd/yyyy hh:mm:ss")
        fw.append(df.format(new java.util.Date()) + "," + size + "," + sortTime+"\r\n")
      } finally fw.close()

    }
}

File: trunk/Documentation/Threading Research/Applications/Go-Multiproc/src/stock/stock.go
package main
/*
*TITLE: 	Stock GO Sorting Application
*AUTOHR: 	THOMAS TRASK
*STARTDATE: 12-5-2012
*
*PURPOSE: 	TO GENERATE A LIST OF RANDOM NUMBERS AND
*			SORT THEM USING THE BUILT-IN GO UTILITIES.  THIS APPLICATION WAS
*			DEVELOPED AS A "CONTROL" BENCHMARK TO DETERMINE IF USING MULTITHREADING
*			/CLUSTER COMPUTING HAS ANY REAL PERFORMANCE BENEFIT.
*
*PROCESS:	THIS PROGRAM CONSISTS OF 2 PARTS: GENERATION, SORT AND
*			VERIFICATION & OUTPUT.
*			GENERATION:	THE APPLICATION GENERETATES A SET OF RANDOM NUMBERS TO SORT.
*			SORT:		THE APPLICATION THEN SORTS THIS DATA SET USING
						THE "sort" GO package.
*			VERIFICATION & OUTPUT: THE SORTED DATA SET IS VERIFIED AS WORKING.
*					ONCE VERIFIED, ALL IMPORTANT PARAMETERS ARE WRITTEN OUT TO
*					AN OUTPUT FILE.
*
*EXECUTION:	There are two ways to run the applicaiton:
*				Using the Go Runtime
*					go run stock.go
*				Compile and Run
*					go build stock.go
*					./stock
*
*INPUT PARAMETERS:
*			-n:		Number of Ints to generate and sort
*			-o:		The name of the output file to write the results to.				
 */
import (
	"flag"
	"fmt"
	"math/rand"
	"os"
	"runtime"
	"sort"
	"time"
	"utility"
)

var _ints int = 100000000

var _threads int = runtime.NumCPU()
var intsFlag *int = flag.Int("n", _ints, "Number of Ints to parse")
var outputFlag *string = flag.String("o", "out.go.stock.csv", "Output File.")

type IntList []int

func main() {

	flag.Parse()

	_ints = *intsFlag

	fmt.Println("Starting Application")

	fmt.Println(fmt.Sprintf("Processing using Parameters: -n %d -o %s", *intsFlag, *outputFlag))

	rand.Seed(time.Now().UTC().UnixNano())

	var m []int = make([]int, _ints)

	fmt.Println("Generating List")

	utility.GenerateList(&m)

	startTime := time.Now()

	fmt.Println("Sorting List")

	sort.Ints(m)

	endTime := time.Since(startTime)
	fmt.Println("Total Elapsed Time:", time.Since(startTime))
	fmt.Println("All Lists Done Sorting")

	fmt.Println("Verifying List is sorted.")

	isSorted := true

	for i := 0; i < len(m)-1; i++ {
		if m[i] > m[i+1] {
			isSorted = false
			i = len(m)
		}
	}

	if isSorted == true {
		fmt.Println("List verified as sorted in", time.Since(startTime))
	} else {
		fmt.Println("****LIST FIALED VERIFICATION. LIST IS NOT SORTED.*********")
	}

	fmt.Println("Writing output")
	writeOutput(fmt.Sprintf("%d,%d\r\n", _ints, endTime), *outputFlag)
}

func writeOutput(out string, file string) {

	if _, err := os.Stat(file); os.IsNotExist(err) {
		os.Create(file)
	}

	f, err := os.OpenFile(file, os.O_APPEND|os.O_WRONLY, 0600)
	if err != nil {
		panic(err)
	}

	defer f.Close()

	if _, err = f.WriteString(out); err != nil {
		panic(err)
	}

}
